bundles/alpha.codegen/src/alpha/codegen/writeC/SystemConverter.xtend

```diff
@@ -0,0 +1,473 @@
+import alpha.model.Variable
+import alpha.model.transformation.Normalize
+import alpha.model.transformation.StandardizeNames
+import fr.irisa.cairn.jnimap.isl.ISLPWQPolynomial
+import java.util.ArrayList
+
+import static extension alpha.codegen.Factory.customExpr
+import static extension alpha.codegen.isl.PolynomialConverter.convert
+import static extension alpha.codegen.writeC.Common.getEvalName
+import static extension alpha.codegen.writeC.Common.getFlagName
+import static extension alpha.model.util.AlphaUtil.copyAE
+import static extension fr.irisa.cairn.jnimap.barvinok.BarvinokBindings.card
+
+/** Converts an Alpha system to the simplified C AST. */
+class SystemConverter {
+	/** The list of all variables allocated by the entry point function which need to be freed before returning. */
+	protected val ArrayList<String> allocatedVariables
+
+	/**
+	 * A counter for the number of statement macros that have been created.
+	 * This is used for determining the names said macros.
+	 */
+	var nextStatementId = 0
+	
+	/**
+	 * Flag to indicate that the C code being output should be compatible
+	 * with the wrapper produced by the older AlphaZ.
+	 */
+	protected val boolean oldAlphaZCompatible
+	
+	/** The instance of the program builder. */
+	protected val ProgramBuilder program
+	
+	/** The system being converted into a C program. */
+	protected val AlphaSystem system
+	
+	/** Protected constructor. */
+	protected new(AlphaSystem system, boolean oldAlphaZCompatible) {
+		this.allocatedVariables = newArrayList
+		this.oldAlphaZCompatible = oldAlphaZCompatible
+		this.program = ProgramBuilder.start
+		this.system = system
+	}
+	
+	/**
+	 * Converts an Alpha system to the simplified C AST.
+	 * Only supports systems with a single body.
+	 */
+	def static convert(AlphaSystem system) {
+		return convert(system, false)
+	}
+	
+	/**
+	 * Converts an Alpha system to the simplified C AST.
+	 * Only supports systems with a single body.
+	 * 
+	 * If requested, the code produced will aim for compatibility with
+	 * the older version of AlphaZ (although with no guarantees).
+	 * This means that the inputs to the entry point will assume
+	 * that memory for the system's inputs and outputs were allocated
+	 * via the bounding box method, as opposed to linearized memory.
+	 * 
+	 * Note: since the newer Alpha language doesn't have typing,
+	 * all indices are assumed to be of type "long",
+	 * and all data values are of type "float".
+	 */
+	def static convert(AlphaSystem system, boolean oldAlphaZCompatible) {
+		// Duplicate the system and then preprocess the duplicate
+		// by applying normalization and name standardization.
+		val duplicate = system.copyAE
+		Normalize.apply(duplicate)
+		StandardizeNames.apply(duplicate)
+		
+		return (new SystemConverter(duplicate, oldAlphaZCompatible)).convertSystem
+	}
+	
+	/**
+	 * Converts an Alpha system to the simplified C AST.
+	 * Only supports systems with a single body.
+	 */
+	def protected convertSystem() {
```

Comment:
This is a good place to start reviewing.

Reply:
Looks really nice overall! I just left a few comments/questions. We can add new issues/tasks as we begin to use it more and more.